lib/ipfs/@types/ipfs/index.d.ts

```diff
@@ -5,188 +5,173 @@ declare module 'ipfs' {
 type Callback<T> = (error: Error, result?: T) => void;
 
 declare class IPFS {
-    constructor(options?: IPFS.Options);
+  constructor (options?: IPFS.Options);
 
-    types: IPFS.Types;
+  types: IPFS.Types;
 
-    init(options: IPFS.InitOptions, callback: Callback<boolean>): void;
-    init(callback: Callback<boolean>): void;
+  init (options: IPFS.InitOptions, callback: Callback<boolean>): void;
+  init (callback: Callback<boolean>): void;
 
-    preStart(callback: Callback<any>): void;
-    start(callback?: Callback<any>): void;
-    stop(callback?: (error?: Error) => void): void;
-    isOnline(): boolean;
+  preStart (callback: Callback<any>): void;
+  start (callback?: Callback<any>): void;
+  stop (callback?: (error?: Error) => void): void;
+  isOnline (): boolean;
 
-    version(options: any, callback: (error: Error, version: IPFS.Version) => void): void ;
-    version(options: any): Promise<IPFS.Version>;
-    version(callback: (error: Error, version: IPFS.Version) => void): void ;
-    version(): Promise<IPFS.Version>;
+  version (options: any, callback: (error: Error, version: IPFS.Version) => void): void ;
+  version (options?: any): Promise<IPFS.Version>;
+  version (callback: (error: Error, version: IPFS.Version) => void): void ;
 
-    id(options: any, callback: (error: Error, version: IPFS.Id) => void): void ;
-    id(options: any): Promise<IPFS.Id>;
-    id(callback: (error: Error, version: IPFS.Id) => void): void ;
-    id(): Promise<IPFS.Id>;
+  id (options: any, callback: (error: Error, version: IPFS.Id) => void): void ;
+  id (options?: any): Promise<IPFS.Id>;
+  id (callback: (error: Error, version: IPFS.Id) => void): void ;
 
-    repo: IPFS.RepoAPI;
-    bootstrap: any;
-    config: any;
-    block: any;
-    object: IPFS.ObjectAPI;
-    dag: IPFS.DagAPI;
-    libp2p: any;
-    swarm: IPFS.SwarmAPI;
-    bitswap: any;
-    pin: IPFS.PinAPI;
+  repo: IPFS.RepoAPI;
+  bootstrap: any;
+  config: any;
+  block: any;
+  object: IPFS.ObjectAPI;
+  dag: IPFS.DagAPI;
+  libp2p: any;
+  swarm: IPFS.SwarmAPI;
+  bitswap: any;
+  pin: IPFS.PinAPI;
 
-    ping(callback: (error: Error) => void): void;
-    ping(): Promise<void>;
+  ping (callback: (error: Error) => void): void;
+  ping (): Promise<void>;
 
-    pubsub: any;
+  pubsub: any;
 
-    on(event: string, callback: () => void): IPFS;
-    on(event: 'error', callback: (error: { message: any }) => void): IPFS;
-    once(event: string, callback: () => void): IPFS;
- 
-    createAddStream(options: any, callback: Callback<any>): void;
-    createAddStream(callback: Callback<any>): void;
+  on (event: string, callback: () => void): IPFS;
+  on (event: 'error', callback: (error: { message: any }) => void): IPFS;
+  once (event: string, callback: () => void): IPFS;
 
-    createPullStream(options: any): any;
+  add (data: IPFS.FileContent, options?: any): AsyncIterator<any>;
```

Comment:
sanity: wondering why does ````add```` return an AsyncIterable<...>? Sounds like it should return a ````Promise<void>```` maybe?

Reply:
It is an iterator because add can return multiple results. It happens when you are writing a nested file structure like `dir1/dir2/file.txt`. It will first return the add result for dir1, then dir2, then file.txt

Reply:
Ahh .. got it.